Data/ByteString/Lazy.hs

```diff
@@ -1131,6 +1132,18 @@ zipWith f (Chunk a as) (Chunk b bs) = go a as b bs
     to _ (Chunk x' xs) y ys            | not (S.null y) = go x' xs y  ys
     to _ (Chunk x' xs) _ (Chunk y' ys)                  = go x' xs y' ys
 
+-- | A specialised version of `zipWith` for the common case of a
+-- simultaneous map over two ByteStrings, to build a 3rd.
+packZipWith :: (Word8 -> Word8 -> Word8) -> ByteString -> ByteString -> ByteString
+packZipWith _ Empty _ = Empty
+packZipWith _ _ Empty = Empty
+packZipWith f (Chunk a@(S.BS _ al) as) (Chunk b@(S.BS _ bl) bs) = Chunk (S.packZipWith f a b) $
+    case compare al bl of
+        LT -> packZipWith f as $ Chunk (S.drop al b) bs
+        EQ -> packZipWith f as bs
+        GT -> packZipWith f (Chunk (S.drop bl a) as) bs
+{-# INLINE packZipWith #-}
+
```

Comment:
How about adding a RULE similar to the one for the strict version?

Reply:
> Given that in future we intend to decomission the rewrite rule, we can probably remove the sentence advertising it already.
- Bodigrim

Don't know how I feel about adding a rule only for it to be decommisioned later, still added the rule to Data.ByteString,Lazy though

Reply:
@sjakobi As discussed in #208, such rule is harmful when the output of `zipWith` is consumed lazily. We cannot drop the existing rule for strict `zipWith` lightly, without further discussion, but from my perspective we should not introduce it for lazy bytestrings.

Reply:
Instead of copying a rule from the stric version, I would rather add 
```
{-# RULES "Specialise zipWith" pack (zipWith f p q) = packZipWith f p q #-} 
```
(both for strict and lazy bytestrings)

Reply:
After reading #208 again I agree. I changed the rewrite rule

Reply:
```
Data/ByteString.hs:1649:1: warning: [-Winline-rule-shadowing]
    Rule "ByteString specialise zipWith" may never fire
      because ‘pack’ might inline first
    Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘pack’
     |
1649 | "ByteString specialise zipWith" forall (f :: Word8 -> Word8 -> Word8) p q .
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

GHC is complaining. I'm not that read up on the inline/noinline pragmas. What needs to be changed?

Reply:
I added `{-# INLINE [1] pack #-}`  to Data.ByteString and Data.ByteString.Lazy since they only resolve to packBytes and I'm sure it would get inlined anyways.

I added `{-# INLINE [1] zipWith #-}` to Data.ByteString.Lazy but I'm not so sure about the performance characteristics of that. Shouldn't be too big of an issue since it just resolves to `go a as b bs`.

Reply:
Thanks for experimenting. I don't mind `{-# INLINE [1] zipWith #-}`, but not quite sure about `{-# INLINE [1] pack #-}`. It seems innocent, but deserves a thorough investigation of the impact, because `pack` is used pretty much everywhere. 

To stop going down the rabbit hole, let's reset back to d93cdd4 and be with it for now. Rules issue can be resolved in a separate PR later. Sorry that it became so laborious.

Reply:
Apologies for the confusion – I had missed the discussion in #208.